rcl/include/rcl/publisher.h

```diff
@@ -0,0 +1,277 @@
+ * \return RCL_RET_OK if the message was published successfully, or
+ *         RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
+ *         RCL_RET_PUBLISHER_INVALID if the publisher is invalid, or
+ *         RCL_RET_ERROR if an unspecified error occurs.
+ */
+RCL_PUBLIC
+rcl_ret_t
+rcl_publish(const rcl_publisher_t * publisher, const void * ros_message);
+
+/// Get the topic name for the publisher.
+/* This function returns the publisher's internal topic name string.
+ * This function can fail, and therefore return NULL, if the:
+ *   - publisher is NULL
+ *   - publisher is invalid (never called init, called fini, or invalid node)
+ *
+ * The returned string is only valid as long as the rcl_publisher_t is valid.
+ * The value of the string may change if the topic name changes, and therefore
+ * copying the string is recommended if this is a concern.
+ *
+ * This function is not thread-safe, and copying the result is not thread-safe.
+ * \TODO(wjwwood): should we provide a thread-safe copy_topic_name?
+ *
+ * \param[in] publisher pointer to the publisher
+ * \return name string if successful, otherwise NULL
+ */
+RCL_PUBLIC
+const char *
+rcl_publisher_get_topic_name(const rcl_publisher_t * publisher);
+
+/// Return the rcl publisher options.
+/* This function returns the publisher's internal options struct.
+ * This function can fail, and therefore return NULL, if the:
+ *   - publisher is NULL
+ *   - publisher is invalid (never called init, called fini, or invalid node)
+ *
+ * The returned struct is only valid as long as the rcl_publisher_t is valid.
+ * The values in the struct may change if the options of the publisher change,
+ * and therefore copying the struct is recommended if this is a concern.
+ *
+ * This function is not thread-safe, and copying the result is not thread-safe.
+ * \TODO(wjwwood): should we provide a thread-safe copy_publisher_options?
+ *
+ * \param[in] publisher pointer to the publisher
+ * \return options struct if successful, otherwise NULL
+ */
+RCL_PUBLIC
+const rcl_publisher_options_t *
+rcl_publisher_get_options(const rcl_publisher_t * publisher);
+
+/// Return the rmw publisher handle.
+/* The handle returned is a pointer to the internally held rmw handle.
+ * This function can fail, and therefore return NULL, if the:
+ *   - publisher is NULL
+ *   - publisher is invalid (never called init, called fini, or invalid node)
+ *
+ * The returned handle is only valid as long as the given publisher is valid.
+ *
+ * \TODO(wjwwood) will the publisher ever need to remove and recreate the rmw
+ *                handle under the hood? Perhaps not now, but if we add the
+ *                ability to reconfigure, alias, or remap topics, and don't
+ *                implement that into rmw, then we should consider the
+ *                implications here. This argument applies to other parts of
+ *                the interface, but I'll just mention it here.
```

Comment:
I removed the TODO and updated the docs to make a note about this in a44c14c.

I added similar notes for node, subscription, and guard condition.

I still have a lingering concern that in the future it may be possible for the handle to be invalidated asynchronously and therefore make using this function always unsafe. I imagine:

- Someone calls this function.
- Externally someone calls a remapping service call which causes the rmw handle to be destroyed and recreated.
- Someone trys to use the result of this function which is now invalid.

I guess it will be up to the user of rcl to make sure the callback for handling remapping or reconfiguring is locked against use of this function?

Having said all that I think it is best to leave it as is (with a note about the volatile nature of the handle) and address it later when we have the need.